### PR TITLE
feat(server): Support CLIENT TRACKING subcommand (2/2)

### PR DESCRIPTION
**Pull request by theyueli**
This is part two that implements the logic which notifies tracking clients by sending invalidation messages. 

Once part 1 is merged, this PR will gets smaller (it includes part 1 currently)

Functional tests have also been included.

### CODE REVIEW

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -1387,11 +1418,11 @@ unsigned Connection::WeakRef::Thread() const {
 }
 
 Connection* Connection::WeakRef::Get() const {
-  DCHECK_EQ(ProactorBase::me()->GetPoolIndex(), int(thread_));
-  // The connection can only be deleted on this thread, so
-  // this pointer is valid until the next suspension.
-  // Note: keeping a shared_ptr doesn't prolong the lifetime because
-  // it doesn't manage the underlying connection. See definition of `self_`.
+  // DCHECK_EQ(ProactorBase::me()->GetPoolIndex(), int(thread_));
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:1424]: This dcheck is here because using the pointer from Get() in another thread is unsafe

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
...
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(const std::string_view& key) {
+  if (client_tracking_map_.find(key) != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = client_tracking_map_[key];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_closed_or_not_tracking = [](const facade::Connection::WeakRef& p) {
+      return (p.IsExpired() || (!p.Get()->IsTrackingOn()));
```
Comment (dranikpg) [inline src/server/db_slice.cc:1389]: Technically, this Get here is unsafe. In practice the the other thread can never be so fast to delete the connection faster than reading a variable...

Better do it from cb, as you drop them below either way

File: src/server/db_slice.cc [hunk 3]
```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
...
+            continue;
+          facade::Connection* conn = it->Get();
+          conn->SendInvalidationMessageAsync({key});
+          return;
+        }
+      };
+      shard_set->pool()->DispatchBrief(std::move(cb));
+    }
+    // remove this key from the tracking table as the key no longer exists
+    client_tracking_map_.erase(key);
```
Comment (dranikpg) [inline src/server/db_slice.cc:1409]: I'm a little confused why this happens. So after the connection got an invalidate it has read the key again to become subscibed again?

File: src/facade/dragonfly_connection.cc [hunk 4]
```diff
@@ -1400,11 +1421,11 @@ unsigned Connection::WeakRef::Thread() const {
...
```
Comment (theyueli) [inline src/facade/dragonfly_connection.cc:1424]: @dranikpg somehow I still need to comment this line although I have been using `IsExpired()`

File: src/server/db_slice.cc [hunk 5]
```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
...
+    auto& client_set = client_tracking_map_[key];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_closed_or_not_tracking = [](const facade::Connection::WeakRef& p) {
+      return (p.IsExpired() || (!p.Get()->IsTrackingOn()));
+    };
+    absl::erase_if(client_set, is_closed_or_not_tracking);
+    DVLOG(2) << "Number of clients left: " << client_set.size();
+
+    if (!client_set.empty()) {
+      auto cb = [key, client_set](unsigned idx, util::ProactorBase*) {
```
Comment (dranikpg) [inline src/server/db_slice.cc:1395]: Better use [client_set = std::move(client_set)]

File: src/server/db_slice.cc [hunk 6]
```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
...
+  for (auto key : keys) {
+    DVLOG(2) << "Inserting client ID " << conn.GetClientId()
+             << " into the tracking client set of key " << key;
+    client_tracking_map_[key].insert(conn);
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(const std::string_view& key) {
+  if (client_tracking_map_.find(key) != client_tracking_map_.end()) {
+    // notify all the clients.
```
Comment (dranikpg) [inline src/server/db_slice.cc:1385]: Some nits: 
You can pass string_view by value
You can store the it = find(key) to avoid reading again with [key]

File: src/server/db_slice.cc [hunk 7]
```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
...
+
+    if (!client_set.empty()) {
+      auto cb = [key, client_set](unsigned idx, util::ProactorBase*) {
+        for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+          if (it->IsExpired())
+            continue;
+          if ((unsigned int)it->Thread() != idx)
+            continue;
+          facade::Connection* conn = it->Get();
+          conn->SendInvalidationMessageAsync({key});
```
Comment (dranikpg) [inline src/server/db_slice.cc:1402]: Conn can be null if it expired, but you check it above and skip it, so its never. Maybe its simpler just to check Get()!=null? 🤔

File: src/server/db_slice.cc (hunk 3 above)
Comment (theyueli) [reply src/server/db_slice.cc:1409]: correct, this is what Redis' implementation is doing.

File: src/server/db_slice.cc (hunk 7 above)
Comment (theyueli) [reply src/server/db_slice.cc:1402]: fixed

File: src/server/db_slice.cc (hunk 5 above)
Comment (theyueli) [reply src/server/db_slice.cc:1395]: ok

File: src/server/db_slice.cc (hunk 6 above)
Comment (theyueli) [reply src/server/db_slice.cc:1385]: fixed

File: src/facade/dragonfly_connection.cc (hunk 1 above)
Comment (theyueli) [reply src/facade/dragonfly_connection.cc:1424]: i see, let me change the use of Get() as you suggested below...

File: src/server/db_slice.cc (hunk 7 above)
Comment (theyueli) [reply src/server/db_slice.cc]: > Get()!=null?

but I still remember in the previous big PR (that has been closed), you suggest use IsExpired() here?

File: src/facade/dragonfly_connection.cc (hunk 1 above)
Comment (theyueli) [reply src/facade/dragonfly_connection.cc]: now I have improved the notification logic, this DCHECK has now come back :-)

File: src/server/db_slice.cc (hunk 7 above)
Comment (dranikpg) [reply src/server/db_slice.cc:1394]: You can use Expire from any thread as you do before disptach brief now

File: src/server/db_slice.cc [hunk 8]
```diff
@@ -1385,27 +1385,16 @@ void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
...
+      for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+        if ((unsigned int)it->Thread() != idx)
+          continue;
+        facade::Connection* conn = it->Get();
+        if ((conn != nullptr) && conn->IsTrackingOn())
+          conn->SendInvalidationMessageAsync({key});
+      }
     };
-    absl::erase_if(client_set, is_closed_or_not_tracking);
-    DVLOG(2) << "Number of clients left: " << client_set.size();
```
Comment (dranikpg) [inline src/server/db_slice.cc]: I thought you do this to reduce the number of traversals (because each thread currently iterates over the whole set)

File: src/server/db_slice.cc [hunk 9]
```diff
@@ -626,7 +634,10 @@ void DbSlice::FlushDbIndexes(const std::vector<DbIndex>& indexes) {
       if (db_ptr && db_ptr->stats.tiered_entries > 0) {
         for (auto it = db_ptr->prime.begin(); it != db_ptr->prime.end(); ++it) {
           if (it->second.IsExternal()) {
+            std::string tmp;
+            std::string_view key = it->first.GetSlice(&tmp);
             PerformDeletion(it, shard_owner(), db_ptr.get());
+            SendInvalidationTrackingMessage(key);
```
Comment (adiholden) [inline src/server/db_slice.cc:640]: why dont you make PerformDeletion function of DbSlice? I dont like it that we need to remember to call SendInvalidationTrackingMessage whenever we call PerformDeletion

File: src/server/server_family.cc [hunk 10]
```diff
@@ -1164,10 +1164,29 @@ string GetPassword() {
   return "";
 }
 
+void ServerFamily::SendInvalidationMessages() const {
+  // send invalidation message (caused by flushdb) to all the clients which
```
Comment (adiholden) [inline src/server/server_family.cc:1168]: how does this work in redis if there are several dbs? 
flush db is not flush all

File: src/server/db_slice.cc (hunk 9 above)
Comment (theyueli) [reply src/server/db_slice.cc:640]: ok, will do.

File: src/server/server_family.cc (hunk 10 above)
Comment (theyueli) [reply src/server/server_family.cc:1168]: > how does this work in redis if there are several dbs? flush db is not flush all

Good question, and Redis does the same. 

image

This behavior actually was in Redis' spec of client side caching, and the reason is simply about lowering implementation complexity:

[https://redis.io/docs/manual/client-side-caching/](https://www.google.com/url?q=https://redis.io/docs/manual/client-side-caching/&sa=D&source=docs&ust=1702327521483711&usg=AOvVaw0QFAUH76qBoZ5hALTYZ_BU)

"There is a single keys namespace, not divided by database numbers. So if a client is caching the key foo in database 2, and some other client changes the value of the key foo in database 3, an invalidation message will still be sent. This way we can ignore database numbers reducing both the memory usage and the implementation complexity."

File: src/server/db_slice.cc (hunk 9 above)
Comment (theyueli) [reply src/server/db_slice.cc]: > why dont you make PerformDeletion function of DbSlice? I dont like it that we need to remember to call SendInvalidationTrackingMessage whenever we call PerformDeletion

done, please take another look. I like this change, saved a lot of code duplication.

File: src/server/server_family.cc [hunk 11]
```diff
@@ -1164,10 +1164,29 @@ string GetPassword() {
...
+  for (auto* listener : listeners_) {
+    listener->TraverseConnections(cb);
+  }
+}
+
 void ServerFamily::FlushDb(CmdArgList args, ConnectionContext* cntx) {
   DCHECK(cntx->transaction);
   Drakarys(cntx->transaction, cntx->transaction->GetDbIndex());
-
+  SendInvalidationMessages();
```
Comment (adiholden) [inline src/server/server_family.cc:1189]: call this from flushall as well

File: src/facade/dragonfly_connection.h [hunk 12]
```diff
@@ -116,6 +116,11 @@ class Connection : public util::Connection {
     util::fb2::BlockingCounter bc;  // Decremented counter when processed
   };
 
+  struct InvalidationMessage {
+    std::string_view key;
```
Comment (adiholden) [inline src/facade/dragonfly_connection.h:120]: I believe you should use a string here and not string_view.
Who is the owner of this string and how do you know it is alive until you send the message?

File: src/server/db_slice.h [hunk 13]
```diff
@@ -338,6 +336,12 @@ class DbSlice {
   // Track keys for the client represented by the the weak reference to its connection.
   void TrackKeys(const facade::Connection::WeakRef&, const ArgSlice&);
 
+  // Send invalidation message to the clients that are tracking the change to a key.
+  void SendInvalidationTrackingMessage(std::string_view key);
```
Comment (adiholden) [inline src/server/db_slice.h:340]: private?

File: src/server/db_slice.cc [hunk 14]
```diff
@@ -1355,4 +1297,85 @@ void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice&
...
+          conn->SendInvalidationMessageAsync({key});
+      }
+    };
+    shard_set->pool()->DispatchBrief(std::move(cb));
+    // remove this key from the tracking table as the key no longer exists
+    client_tracking_map_.erase(key);
+  }
+}
+
+void DbSlice::PerformDeletion(PrimeIterator del_it, ExpireIterator exp_it, EngineShard* shard,
```
Comment (adiholden) [inline src/server/db_slice.cc:1320]: Its hard to review with github. Have you done any changes in the function instead of moving it to be member of the class?

File: src/server/db_slice.cc [hunk 15]
```diff
@@ -1233,6 +1172,8 @@ size_t DbSlice::EvictObjects(size_t memory_to_free, PrimeIterator it, DbTable* t
     return current < used_memory_start ? used_memory_start - current : 0;
   };
 
+  string_view key;
```
Comment (adiholden) [inline src/server/db_slice.cc:1175]: redundant

File: src/server/db_slice.cc [hunk 16]
```diff
@@ -994,18 +936,15 @@ pair<PrimeIterator, ExpireIterator> DbSlice::ExpireIfNeeded(const Context& cntx,
...
 
   // Replicate expiry
   if (auto journal = owner_->journal(); journal) {
-    tmp_key = it->first.GetSlice(&tmp_key_buf);
     RecordExpiry(cntx.db_index, tmp_key);
   }
 
   auto obj_type = it->second.ObjType();
   if (doc_del_cb_ && (obj_type == OBJ_JSON || obj_type == OBJ_HASH)) {
-    if (tmp_key.empty())
```
Comment (adiholden) [inline src/server/db_slice.cc:1007]: I believe you can revert the changes in this function

File: src/server/db_slice.cc (hunk 14 above)
Comment (theyueli) [reply src/server/db_slice.cc:1320]: no change in the function, but added invalidation notification at the end of the PerformDeletion implementation.

File: src/server/db_slice.h (hunk 13 above)
Comment (theyueli) [reply src/server/db_slice.h:340]: yes, fixed.

File: src/server/server_family_test.cc [hunk 17]
```diff
@@ -200,4 +200,52 @@ TEST_F(ServerFamilyTest, ClientPause) {
...
+  Run({"SET", "BAR", "1"});
+  Run({"GET", "BAR"});
+  pp_->at(1)->Await([&] { return Run({"DEL", "BAR"}); });
+  EXPECT_EQ(GetInvalidationMessage("IO0", 2).key, "BAR");
+
+  // case 7. test multi command
+  Run({"MGET", "X", "Y", "Z"});
+  pp_->at(1)->Await([&] { return Run({"MSET", "X", "1", "Y", "2"}); });
+  EXPECT_EQ(GetInvalidationMessage("IO0", 3).key, "X");
+  EXPECT_EQ(GetInvalidationMessage("IO0", 4).key, "Y");
```
Comment (dranikpg) [inline src/server/server_family_test.cc:242]: I'd make the test a little more sophisticated. Let's try multiple keys on all shards, so that we don't even know the order

File: src/facade/dragonfly_connection.cc [hunk 18]
```diff
@@ -288,6 +292,19 @@ void Connection::DispatchOperations::operator()(CheckpointMessage msg) {
...
+void Connection::DispatchOperations::operator()(const InvalidationMessage& msg) {
+  RedisReplyBuilder* rbuilder = (RedisReplyBuilder*)builder;
+  DCHECK(rbuilder->IsResp3());
+  rbuilder->StartCollection(2, facade::RedisReplyBuilder::CollectionType::PUSH);
+  rbuilder->SendBulkString("invalidate");
+  if (msg.invalidate_due_to_flush) {
+    rbuilder->SendNull();
+  } else {
+    std::vector<string_view> keys{msg.key};
+    rbuilder->SendStringArr(keys);
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:304]:
```c++
string_view[] keys = {msg.key}
SendStringArr(keys)
```
does the same without allocations

**Review by dranikpg — COMMENTED** [review]
The testing approach is the same as I had in mind. Please note there are still some unresolved comments from Adi

**Comment by theyueli** [comment]
> The testing approach is the same as I had in mind. Please note there are still some unresolved comments from Adi

yes, i plan to add more tests now after verifying the mock methods with you guys.

File: src/server/server_family.cc (hunk 11 above)
Comment (theyueli) [reply src/server/server_family.cc:1189]: fixed

File: src/server/server_family_test.cc [hunk 19]
```diff
@@ -200,4 +200,52 @@ TEST_F(ServerFamilyTest, ClientPause) {
   EXPECT_GT((absl::Now() - start), absl::Milliseconds(50));
 }
 
+TEST_F(ServerFamilyTest, ClientTracking) {
```
Comment (adiholden) [inline src/server/server_family_test.cc:203]: I would create different unit test for different test cases.
I think that it is great that we have this in unit tests with mocking but we must add pytest to make sure this command works well to check the functionality without the mocking

Comment (theyueli) [reply src/server/server_family_test.cc:203]: that's fine with me too. Currently i see different styles mixed in our code: sometimes all cases went into one function (e.g. ClientPause here, and many other redis commands' tests), sometimes separated smaller cases (e.g. SlowLog here)

File: src/server/server_family_test.cc (hunk 17 above)
Comment (theyueli) [reply src/server/server_family_test.cc:242]: even with this test, the order of arrival is different now, btw, is there a convenient test function that checks the completeness disregarding the order?

File: src/server/db_slice.cc (hunk 15 above)
Comment (theyueli) [reply src/server/db_slice.cc:1175]: fixed

File: src/server/server_family_test.cc (hunk 17 above)
Comment (theyueli) [reply src/server/server_family_test.cc]: added a few more keys, how does this look?

File: src/facade/dragonfly_connection.cc (hunk 18 above)
Comment (theyueli) [reply src/facade/dragonfly_connection.cc:304]: This doesn't compile

```
/home/yli/git/dragonfly-rebase/src/facade/dragonfly_connection.cc:303:21: error: structured binding declaration cannot have type ‘std::string_view’ {aka ‘std::basic_string_view<char>’}
  303 |     std::string_view[] keys = {msg.key};
      |                     ^~
/home/yli/git/dragonfly-rebase/src/facade/dragonfly_connection.cc:303:21: note: type must be cv-qualified ‘auto’ or reference to cv-qualified ‘auto’
/home/yli/git/dragonfly-rebase/src/facade/dragonfly_connection.cc:303:21: error: empty structured binding declaration
/home/yli/git/dragonfly-rebase/src/facade/dragonfly_connection.cc:303:24: error: expected initializer before ‘keys’
  303 |     std::string_view[] keys = {msg.key};
      |                        ^~~~
/home/yli/git/dragonfly-rebase/src/facade/dragonfly_connection.cc:304:29: error: ‘keys’ was not declared in this scope
  304 |     rbuilder->SendStringArr(keys);
      |                             ^~~~
```

Comment (chakaz) [reply src/facade/dragonfly_connection.cc:304]: try `string_view keys[] = {msg.key};`
It thinks you're using structured binding, using the brackets on the left side is C#/Java syntax

Comment (theyueli) [reply src/facade/dragonfly_connection.cc:304]: this works well, thanks!

File: src/facade/dragonfly_connection.h (hunk 12 above)
Comment (theyueli) [reply src/facade/dragonfly_connection.h:120]: fixed.

**Comment by theyueli** [comment]
all the comments have been addressed. Please take another look. @dranikpg @adiholden @chakaz

File: src/server/db_slice.cc [hunk 20]
```diff
@@ -1049,18 +1051,15 @@ DbSlice::ItAndExp DbSlice::ExpireIfNeeded(const Context& cntx, PrimeIterator it)
     return {it, expire_it};
 
   string tmp_key_buf;
-  string_view tmp_key;
+  string_view tmp_key = it->first.GetSlice(&tmp_key_buf);
```
Comment (adiholden) [inline src/server/db_slice.cc:1054]: please revert the changes here to be as before. you did not changed this function

Comment (theyueli) [reply src/server/db_slice.cc:1054]: done

File: src/server/server_family.cc [hunk 21]
```diff
@@ -1172,10 +1172,29 @@ string GetPassword() {
...
+    }
+  };
+  for (auto* listener : listeners_) {
+    listener->TraverseConnections(cb);
+  }
+}
+
 void ServerFamily::FlushDb(CmdArgList args, ConnectionContext* cntx) {
   DCHECK(cntx->transaction);
   Drakarys(cntx->transaction, cntx->transaction->GetDbIndex());
```
Comment (adiholden) [inline src/server/server_family.cc:1196]: I think we should clear client_tracking_map_ in flush

File: src/server/db_slice.cc [hunk 22]
```diff
@@ -1410,4 +1413,87 @@ void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice&
...
+  if (ClusterConfig::IsEnabled()) {
+    SlotId sid = ClusterConfig::KeySlot(key);
+    table->slots_stats[sid].key_count -= 1;
+  }
+
+  table->prime.Erase(del_it);
+  SendInvalidationTrackingMessage(key);
+}
+
+void DbSlice::PerformDeletion(PrimeIterator del_it, EngineShard* shard, DbTable* table) {
```
Comment (adiholden) [inline src/server/db_slice.cc:1489]: After reviewing this more times and reviewing other PR in this file I now understand that actually PerformDeletion should not be a member of DbSlice as I asked before. Rather we should have in DbSlice and function DeleteInternal which will write to journal, do the doc_del_cb_ call, SendInvalidationTrackingMessage
This function will be called from all the flows except the flush flows in which we dont need to do this changes and there we will call PerformDeletion which will be static function.

I believe this will fix several bugs we have that we forget to call the doc_del_cb_ and the journal change writes

File: src/server/server_family.cc (hunk 21 above)
Comment (theyueli) [reply src/server/server_family.cc:1196]: this is not needed as Drakarys will finally end up calling PerformDeletion for each key, which will get the tracking table removed at the end.

File: src/server/db_slice.cc (hunk 22 above)
Comment (theyueli) [reply src/server/db_slice.cc:1489]: This kind of refactoring is out of the scope for this PR. We will do it in another one once this gets merged.

File: src/server/server_family.cc (hunk 21 above)
Comment (adiholden) [reply src/server/server_family.cc:1196]: FushDb calls PerformDeletion only for external keys (ssd).
If we would have called PerformDeletion for all keys in flushdb the client would get the invalidation message twice, one for all keys and one for each key in the database.
If we will have client tracking with tiering I believe the behaviour of flushall will be wrong. This is why I wrote above to get the invalidation message out of PerformDeletion that should just free the data and update table counters

File: src/server/db_slice.cc (hunk 22 above)
Comment (adiholden) [reply src/server/db_slice.cc:1494]: Yes the refactoring can be done in a seprate PR. But as I wrote in another comment I believe that sending Invalidation message from PerformDeletion is incorrect for the flow of flush all.

**Comment by adiholden** [comment]
@theyueli you are going to commit the pytests to this PR correct?

File: src/server/server_family.cc (hunk 21 above)
Comment (theyueli) [reply src/server/server_family.cc:1196]: i see what you meant, I just added another commit which clear client_tracking_map in FlushDB of DbSlice:

https://github.com/dragonflydb/dragonfly/pull/2280/commits/287c3f54f552e4a426c23b6c1689901107b8ab6d

This shall fix both problem:
1) FlushDB and FlushAll of ServerFamily will end up clearing the client tracking map
2) For SSD tiering, since the map has been cleared before `PerformDeletion` gets called, notification will not be sent out twice.

File: src/server/db_slice.cc (hunk 22 above)
Comment (theyueli) [reply src/server/db_slice.cc:1497]: OK, please see if my reply above will resolve this issue as well.

**Comment by theyueli** [comment]
> @theyueli you are going to commit the pytests to this PR correct?

Probably in a separate PR.

**Review by adiholden — APPROVED** [review]
